### DISCUSSION TITLE
FreeBSD setup fails with "Unable to update repository FreeBSD"

### DISCUSSION THREAD

**Discussion by madebr** [cross-platform-actions/action#90] [General]
Hello!

Since today, our ci fails to setup FreeBSD with the following message:
```
Newer FreeBSD version for package zziplib:
To ignore this error set IGNORE_OSVERSION=yes
- package: 1303001
- running kernel: 1302001
Ignore the mismatch and continue? [y/N]: pkg: repository FreeBSD contains packages for wrong OS version: FreeBSD:13:amd64
Processing entries... done
Unable to update repository FreeBSD
Error updating repositories!
```
Does anyone know how to fix this?
https://github.com/libsdl-org/SDL/actions/runs/9881928988/job/27293745437#step:3:115

Updating to latest cross-platform-actions/action, NetBSD and FreeBSD to the latest gives different problems.
A fix for those would be appreciated as well :)

FreeBSD 14.1: https://github.com/madebr/SDL/actions/runs/9882498559/job/27295474947#step:3:57
NetBSD 10: https://github.com/madebr/SDL/actions/runs/9882498559/job/27295475506#step:3:39

**Comment by jacob-carlborg** [comment]
> Since today, our ci fails to setup FreeBSD with the following message:
> 
> ```
> Newer FreeBSD version for package zziplib:
> To ignore this error set IGNORE_OSVERSION=yes
> - package: 1303001
> - running kernel: 1302001
> Ignore the mismatch and continue? [y/N]: pkg: repository FreeBSD contains packages for wrong OS version: FreeBSD:13:amd64
> Processing entries... done
> Unable to update repository FreeBSD
> Error updating repositories!
> ```
> 
> Does anyone know how to fix this? [libsdl-org/SDL/actions/runs/9881928988/job/27293745437#step:3:115](https://github.com/libsdl-org/SDL/actions/runs/9881928988/job/27293745437#step:3:115)

I'm not an expert on FreeBSD, but as far as I understand, this is how it works: for each major version of FreeBSD there's a package repository. This repository is for a given ABI version, i.e. the same as the major version. The major version (13) is supported for five years, but the point version (13.2) is only supported for three months [1]. The package repository only contains packages compiled for the most recent point version. But as far as I know all point versions are binary compatible. If you look here [2], you can see that for FreeBSD 13, the currently supported point version is 13.3. The error message says: `package: 1303001`, which I assume means trying to install a package built for FreeBSD 13.3. Then `running kernel: 1302001`, which seems to be the version of the currently running kernel, i.e. your running FreeBSD 13.2.

Unfortunately this is not something this GitHub action can do anything about. Upstream breaks its users. You have a couple of options:

1. Start using FreeBSD 13.3 (this is the best alternative)
2. If you can't update to 13.3, try setting the environment variable that is mentioned in the error message: `IGNORE_OSVERSION=yes`. It probably will work since the point versions are supposed to be binary compatible. But there is probably no guarantee that it will work.
3. Try to find a package mirror that is has the packages compiled for 13.2.

[1] https://www.freebsd.org/security/#model
[2] https://www.freebsd.org/security/#sup

**Comment by jacob-carlborg** [comment]
> NetBSD 10: [madebr/SDL/actions/runs/9882498559/job/27295475506#step:3:39](https://github.com/madebr/SDL/actions/runs/9882498559/job/27295475506#step:3:39)

NetBSD doesn't work because it's looking for version `10` instead of `10.0`. I had a look the your workflow. You have specified `10.0` as a float instead of as a string. This is a quirk of YAML. YAML tries to figure out the type of a value based on the value itself. So if you write `bar: foo`, `foo` will be interpreted as a string. If you write `bar: 4`, `4` will be interpreted as an integer. If you write `bar: 5.3`, `5.3` will be interpreted as a float. When it comes to floats and when the fractional is `0`, it has no meaning, `10.0 == 10`. That means YAML will drop the fractional part and the GitHub action only receives `10`. This is also mentioned in the documentation below the inputs table [2].

To fix this, make the value an explicit string by adding quotes: `'10.0'`. I recommend always specifying the version in quotes to avoid this issue in the future.

[1] https://github.com/madebr/SDL/commit/eea66b384a8ad611e533176ebb1b14711baf9490#diff-ef1bf491dca9f19ea103f67e323142f68c9f7cf79c699282d0ed931ee3107a83R23

[2] https://github.com/cross-platform-actions/action?tab=readme-ov-file#inputs

**Comment by jacob-carlborg** [comment]
> FreeBSD 14.1: [madebr/SDL/actions/runs/9882498559/job/27295474947#step:3:57](https://github.com/madebr/SDL/actions/runs/9882498559/job/27295474947#step:3:57)

Support for FreeBSD 14.1 has not been released yet, it's only in master so far. The readme is always for master.

**Reply by jacob-carlborg** [reply]
I released a new version now, where support for FreeBSD 14.1 is available: https://github.com/cross-platform-actions/action/releases/tag/v0.25.0.

**Comment by madebr** [comment]
Thanks Jacob for the fast and correct feedback!
We use your action to verify SDL on the latest BSDs.

If they don't worry about breaking backwards compatibility, then porting our internals will be much easier if we don't have to worry about older revisions.

**Reply by jacob-carlborg** [reply]
> If they don't worry about breaking backwards compatibility, then porting our internals will be much easier if we don't have to worry about older revisions.

They do care about backwards compatibility. Each major version is supported for five years. It's the point versions that are only supported for a short period, three months.
